lib/gpw/client.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module GPW
+  class Client
+    def fetch_companies_list
+      response = fetch_gpw_stock_list
+      @companies = parse_data(response)
+      @companies.each { |company| Company.find_or_create_by!(company) }
```

Comment:
1. Why do you need instance variable here?
2. This method is called `fetch_companies_list`, but it does more than that. It stores fetched companies into the database. So, there is a mismatch between the name and actual behaviour at the moment.
  Personally, I would keep the method name, but removed the code that changes records in the database (see point 4).
3. I think there is an implicit assumption here that your db table with companies has the columns named the same way as keys in GPW's API response. What if the GPW would change the API? It would be nice to somehow decouple these 2 things from each other:

    ```ruby
    companies.each do |company|
      Company.find_or_create_by!(
        symbol: company.symbol,
        # ...,
    end
    ```
    
    This way you could name your columns in a different way than the keys in GPW's response body. You may think that this idea requires adding a lot of additional code that repeats the same names. It is true now, but you may want to add other data providers (other than GPW) in the future, and their API would have a different naming or API schema. I think this shows us that the names in your database and the names in GPW may look that same but there should be considered as a two different things.
    You could say that until they do not change the API or until you need to add another provider, it is simpler to keep the code as it is. This would be fine, if you only remember to change the code accordingly at that point. But I would change the code right away in this case.

4. Personally, I would keep `GPW::Client` totally unconscious of any knowledge about the models from your database.In my opinion, it should be a class that could be reused in a different application that does not use any database, etc. This way, your code would be less interconnected and easier to maintain later.
  To achieve this, `fetch_companies_list` could simply return companies as a primitive types such as list of hashes. In the other words, you would need to move this line back to the controller.
  If you want to have something more sophisticated, you could create a class that represents a company fetched from GPW and define it as `GPW::Company`.

Reply:
Everything looks clear, at point 3 I was thinking little bit, but you are right, there is an implicit assumption, the key name in the response may suggest wrong way in the future (I can try to change to much in the future in my code or someone else can). But solution is easy, because this symbol has exact name ISIN, so I will change it.